File: src/spec/resources/testprojects/csharptest/src/PetstoreClient/Auth/OAuth/OAuth2ClientCredentialsAuthenticator.cs
#pragma warning disable IDE0290 // Use primary constructor

namespace PetstoreClient.Auth.OAuth;

/// <summary>
/// Authenticator for the OAuth2 Client Credentials flow.
/// </summary>
public sealed class OAuth2ClientCredentialsAuthenticator : BaseAuthenticator
{
    private readonly string _host;
    private readonly string _clientId;
    private readonly string _clientSecret;
    private readonly Uri _tokenUrl;
    private readonly string[] _scopes;
    private readonly OAuth2TokenManager _tokenManager = new();

    public OAuth2ClientCredentialsAuthenticator(
        string host,
        string clientId,
        string clientSecret,
        Uri tokenUrl,
        string[] scopes
    )
    {
        _host = host;
        _clientId = clientId;
        _clientSecret = clientSecret;
        _tokenUrl = tokenUrl;
        _scopes = [.. scopes];
    }

    /// <inheritdoc/>
    public override string GetHost()
    {
        return _host;
    }

    /// <inheritdoc/>
    public override Dictionary<string, string> GetAuthHeaders()
    {
        Dictionary<string, string> parameters = new()
        {
            ["grant_type"] = "client_credentials",
            ["client_id"] = _clientId,
            ["client_secret"] = _clientSecret,
        };
        if (_scopes.Length > 0)
        {
            parameters["scope"] = string.Join(" ", _scopes);
        }

        string token = _tokenManager
            .GetAccessTokenAsync(_tokenUrl, parameters)
            .GetAwaiter()
            .GetResult();
        return new() { ["Authorization"] = "Bearer " + token };
    }
}

File: src/spec/resources/testprojects/nodetest/src/auth/oauth/oauth2-client-credentials-authenticator.ts
import { BaseAuthenticator } from '../base-authenticator.js';
import { OAuth2TokenManager } from './oauth2-token-manager.js';

/**
 * Authenticator for the OAuth2 Client Credentials flow.
 */
export class OAuth2ClientCredentialsAuthenticator extends BaseAuthenticator {
  private readonly host: string;
  private readonly clientId: string;
  private readonly clientSecret: string;
  private readonly tokenUrl: string;
  private readonly scopes: readonly string[];
  private readonly tokenManager: OAuth2TokenManager;

  constructor(host: string, clientId: string, clientSecret: string, tokenUrl: string, scopes: string[]) {
    super();
    this.host = host;
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.tokenUrl = tokenUrl;
    this.scopes = Object.freeze([...scopes]);
    this.tokenManager = new OAuth2TokenManager();
  }

  getHost(): string {
    return this.host;
  }

  getAuthHeaders(): Record<string, string> {
    throw new Error('Use getAuthHeadersAsync() instead');
  }

  async getAuthHeadersAsync(): Promise<Record<string, string>> {
    const params: Record<string, string> = {
      grant_type: 'client_credentials',
      client_id: this.clientId,
      client_secret: this.clientSecret
    };
    if (this.scopes.length > 0) {
      params.scope = this.scopes.join(' ');
    }
    const token = await this.tokenManager.getAccessToken(this.tokenUrl, params);
    return { Authorization: `Bearer ${token}` };
  }
}
